src/combinatorics/simple_graph/handshake.lean

```diff
@@ -0,0 +1,254 @@
+λ ee, ⟨v, ee, ee.property.1, ee.property.2⟩
+
+lemma dart_from_incidence_set_inj (v : V) : function.injective (G.dart_from_incidence_set v) :=
+by { intros e₁ e₂, dsimp only [dart_from_incidence_set, dart_equiv_sigma],
+     simp only [true_and, equiv.coe_fn_symm_mk, eq_self_iff_true], exact subtype.ext,  }
+
+section degree_sum
+variables [fintype V] [decidable_eq V] [decidable_rel G.adj]
+
+lemma dart_vert_fiber_card_eq_degree (v : V) :
+  (univ.filter (λ d : G.dart, d.vert = v)).card = G.degree v :=
+begin
+  have hh := card_image_of_injective univ (G.dart_from_incidence_set_inj v),
+  rw [finset.card_univ, card_incidence_set_eq_degree] at hh,
+  convert hh,
+  ext d,
+  simp only [mem_image, true_and, mem_filter, set_coe.exists, mem_univ, exists_prop_of_true],
+  split,
+  { rintro rfl, use [d.edge, ⟨d.edge_mem, d.vert_mem⟩],
+    apply dart.ext; simp [dart_from_incidence_set, dart.vert, dart.edge], },
+  { rintro ⟨e, he, rfl⟩, simp [dart_from_incidence_set, dart.vert] },
+end
+
+lemma dart_card_eq_sum_degrees : fintype.card G.dart = ∑ v, G.degree v :=
+begin
+  convert_to (univ : finset G.dart).card = _,
+  rw @card_eq_sum_card_fiberwise _ _ (by assumption) dart.vert _ univ (λ d h, mem_univ _),
+  simp [G.dart_vert_fiber_card_eq_degree],
+end
+
+lemma dart_edge_fiber (d : G.dart) :
+  (univ.filter (λ (d' : G.dart), d'.edge = d.edge)) = {d, d.rev} :=
+begin
+  ext d',
+  simp only [true_and, mem_filter, mem_insert, mem_univ, mem_singleton],
+  exact dart_edge_eq_iff d' d,
+end
+
+lemma dart_edge_fiber_card (e : sym2 V) (h : e ∈ G.edge_set) :
+  (univ.filter (λ (d : G.dart), d.edge = e)).card = 2 :=
+begin
+  refine quotient.ind (λ p h, _) e h, cases p with v w,
+  let d : G.dart := ⟨v, ⟦(v, w)⟧, h, sym2.mk_has_mem _ _⟩,
+  convert_to _ = finset.card {d, d.rev},
+  { rw [card_insert_of_not_mem, card_singleton], simp [d.neq_rev], },
+  congr, apply G.dart_edge_fiber d,
+end
+
+lemma dart_card_eq_twice_card_edges : fintype.card G.dart = 2 * G.edge_finset.card :=
+begin
+  convert_to (univ : finset G.dart).card = _,
+  rw @card_eq_sum_card_fiberwise _ _ _ dart.edge _ G.edge_finset
+    (λ d h, by { rw mem_edge_finset, apply dart.edge_mem }),
+  rw @sum_congr _ _ G.edge_finset G.edge_finset
+    (λ e, (univ.filter (λ (x : G.dart), dart.edge x = e)).card) (λ e, 2) _ rfl,
+  simp [mul_comm],
+  intros e h, rw mem_edge_finset at h,
+  exact G.dart_edge_fiber_card e h,
+end
+
+/-- The degree-sum formula.  This is also known as the handshaking lemma, which might
+more specifically refer to `simple_graph.card_odd_degree_vertices_is_even`. -/
+theorem sum_degrees_eq_twice_card_edges : ∑ v, G.degree v = 2 * G.edge_finset.card :=
+G.dart_card_eq_sum_degrees.symm.trans G.dart_card_eq_twice_card_edges
+
+end degree_sum
+
+
+section TODO_move
+
+lemma zmod_eq_zero_iff_even (n : ℕ) : (n : zmod 2) = 0 ↔ even n :=
+begin
+  change (n : zmod 2) = ((0 : ℕ) : zmod 2) ↔ _,
+  rw [zmod.eq_iff_modeq_nat, nat.even_iff],
+  trivial,
+end
```

Comment:
I can golf this to
```suggestion
(char_p.cast_eq_zero_iff (zmod 2) 2 n).trans even_iff_two_dvd.symm
```
But I think my mathlib might be out of date

Reply:
Worked in mine, so I change it